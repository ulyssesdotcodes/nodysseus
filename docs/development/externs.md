# Externs

## Outputs

Make sure to add output to extern `outputs` field if it's an unusual output otherwise the extern doesn't return it.

Example: tried to add a metadata output to `return` but forgot to update the outputs field, so run_extern thought `return` couldn't handle metadata


