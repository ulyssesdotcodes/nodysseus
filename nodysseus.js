import { json, jsonParseLinter, jsonLanguage } from "@codemirror/lang-json";
import { javascript, javascriptLanguage } from "@codemirror/lang-javascript";
import {keymap, highlightSpecialChars, drawSelection, highlightActiveLine, EditorView} from "@codemirror/view";
import {EditorState, StateEffect, Compartment} from "@codemirror/state";
import {history, historyKeymap} from "@codemirror/history";
import {foldGutter, foldKeymap} from "@codemirror/fold";
import {indentOnInput} from "@codemirror/language";
import {lineNumbers, highlightActiveLineGutter} from "@codemirror/gutter";
import {cursorLineBoundaryBackward, standardKeymap} from "@codemirror/commands";
import {bracketMatching} from "@codemirror/matchbrackets";
import {closeBrackets, closeBracketsKeymap} from "@codemirror/closebrackets";
import {searchKeymap, highlightSelectionMatches} from "@codemirror/search";
import {autocompletion, completionKeymap} from "@codemirror/autocomplete";
import {commentKeymap} from "@codemirror/comment";
import {rectangularSelection} from "@codemirror/rectangular-selection";
import {defaultHighlightStyle} from "@codemirror/highlight";
import {lintKeymap, linter} from "@codemirror/lint";
import {language} from "@codemirror/language";


import * as R from 'ramda';
import { h, app, text, memo } from "hyperapp"

import DEFAULT_GRAPH from "./default.nodysseus.json"
import { times } from "lodash";

const cm = {javascript, json, searchKeymap, keymap, highlightSpecialChars, 
	drawSelection, highlightActiveLine, EditorState, EditorView, StateEffect, jsonParseLinter,
	history, historyKeymap, foldGutter, foldKeymap, indentOnInput, lineNumbers, highlightActiveLineGutter,
	standardKeymap, bracketMatching, closeBrackets, closeBracketsKeymap, searchKeymap, highlightSelectionMatches,
	autocompletion, completionKeymap, commentKeymap, rectangularSelection, defaultHighlightStyle, lintKeymap, linter
};

// helpers

const queue = (init) => ({
	arr: Array.isArray(init) ? init : [init],
	push: function(v){ this.arr.unshift(v) },
	[Symbol.iterator]: function(){
		return {
			next: () => ({ done: this.arr.length === 0, value: this.arr.length === 0 ? undefined : this.arr.pop() })
		}
	}
})

const reduce = (fn, acc, it) => {
	for(const n of it) {
		acc = fn(acc, n);
	}

	return acc;
}

const map = function* (it, fn) {
	for(const n of it) {
		yield fn(n);
	}
}

// in place `over` of an array index
const overIdx = i => fn => arr => (arr[i] = fn(arr[i]), arr);

const nestedPropertyIterator = (val, prop) => ({
	level: val,
	[Symbol.iterator]: function(){
		return {
			next: () => ({ done: this.level.hasOwnProperty(prop), value: (this.level = this.level[prop], this.level) })
		}
	}
})

window.AsyncFunction = Object.getPrototypeOf(async function(){}).constructor;

class CopaneError extends Error {
	constructor(tag, message) {
		super(`${tag}: ${message}`);
	}
}

const rethrow = (tag) => (error) => {
	if (error instanceof CopaneError) {
		throw error;
	} else {
		throw new CopaneError(tag, error.message);
	}
}

const editorEl = document.getElementById("text_editor");
const graphEl = document.getElementById("node_editor");

const languageConf = new Compartment();

const autoLanguage = EditorState.transactionExtender.of(tr => {
  if (!tr.docChanged) return null
  let docIsJSON = true;

	try{
		JSON.parse(tr.newDoc.toString());
	} catch (err){
		docIsJSON = false;
	}

  let stateIsJSON = tr.startState.facet(language) == jsonLanguage;
  if (docIsJSON == stateIsJSON) return null
  return {
    effects: languageConf.reconfigure(docIsJSON ? [cm.linter(cm.jsonParseLinter()), json()] : javascript())
  }
})

const editorState  = cm.EditorState.create({
	extensions: [
		languageConf.of([cm.linter(cm.jsonParseLinter()), json()]),
		autoLanguage,
		cm.lineNumbers(),
		cm.highlightActiveLineGutter(),
		cm.highlightSpecialChars(),
		cm.history(),
		cm.foldGutter(),
		cm.drawSelection(),
		cm.EditorState.allowMultipleSelections.of(true),
		cm.indentOnInput(),
		cm.defaultHighlightStyle.fallback,
		cm.bracketMatching(),
		cm.closeBrackets(),
		cm.autocompletion(),
		cm.rectangularSelection(),
		cm.highlightActiveLine(),
		cm.highlightSelectionMatches(),
		cm.keymap.of([
			...cm.closeBracketsKeymap,
			...cm.standardKeymap,
			...cm.searchKeymap,
			...cm.historyKeymap,
			...cm.foldKeymap,
			...cm.commentKeymap,
			...cm.completionKeymap,
			...cm.lintKeymap
		])
	]
});
			

const editorView = new cm.EditorView({ 
	state: editorState, 
	parent: editorEl
});

const compile = (node) => node.script ? new Function('lib', 'self', node.script)(lib, node) : default_fn(lib, node);

// Note: heavy use of comma operator https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator
const unpackTypes = (node_types, node) => {
	let ty = node.type;
	const result = Object.assign({}, node);
	while(node_types[ty]) {
		Object.assign(result, node_types[ty], result);
		ty = node_types[ty].type;
	}
	return result;
}

const compileNodes = (node_types, nodes) => Object.fromEntries(
	Object.entries(nodes ?? {})
		.map(overIdx(1)(n => Object.assign({}, {node_types}, unpackTypes(node_types, n), n)))
		.map(overIdx(1)(compile))
);

const createEdgeFns = (edges, from, fns) => (edges ?? [])
	.filter(e => e.from === from)
	.map(edge => Object.assign({}, edge, {fn: fns[edge.to]}))

const default_fn = (lib, self) => {
	const node_fns = compileNodes(self.node_types, self.nodes);

	return (state, input) => { 
		const edge_queue = queue(createEdgeFns(self.edges, 'in', node_fns));

		state.set('in', input);

		return reduce((outputs, run_edge) => {
			const next = Object.assign({}, outputs);

			if(run_edge.to === "out") {
				outputs['out'] = outputs[run_edge.from];
				return outputs;
			}

			try {
				const result = run_edge.fn(
					state.get(run_edge.to) ?? state.set(run_edge.to, new Map()).get(run_edge.to),
					run_edge.as ? Object.fromEntries([[run_edge.as, outputs[run_edge.from]]]) : outputs[run_edge.from]
				);

				next[run_edge.to] = result ?? next[run_edge.to];

				if(next[run_edge.to] !== undefined) {
					// side effect add to queue
					createEdgeFns(self.edges, run_edge.to, node_fns)
						.forEach(e => edge_queue.push(e))

					return next;
				}
			} catch(err) {
				console.log("error in " + run_edge.to);
				console.log(run_edge);
				console.log(outputs[run_edge.from]);
				console.log(state.get(run_edge.to));
				console.error(err);
			}

			// don't update if the new value is undefined
			return outputs;
		}, {in: input}, edge_queue).out;
	}
} 

const aggregate_fn = (lib, self) => {
	const node_fns = lib.no.compileNodes(self.node_types, self.nodes); 
	const edge_queue = queue(createEdgeFns(self.edges, 'in', node_fns));
	const output_order = reduce((order, edge) => {
		const has_edge = order.has(edge.to);

		if (!has_edge) {
			createEdgeFns(self.edges, edge.to, node_fns)
				.forEach(e => edge_queue.push(e));
			return order.set(edge.to, edge);
		}

		return order
	}, new Map(), edge_queue);

	return (state, input) => new Map([...map(output_order.entries(), ([k, run_edge]) => [k, run_edge.fn(state.has(k) ? state.get(k) : state.set(k, new Map()).get(k), input)])]);
}

const h_fn = (lib, self) => (state, input) => haState => (console.log(input), lib.ha.h(
		self.dom_type, // change to input
		input.attrs ? input.attrs(haState) : {}, 
		input.children 
			?  [...lib.iter.map(
				input.children.entries(), 
				([k, v]) => v(haState[k]))]
			: []));

// returns a state, input => output

const lib = { ha: { h, app, text, memo},  cm, iter: {reduce, map}, no: { compileNodes, default_fn, aggregate_fn, h_fn }, R };

// sort edges by distance from in
// const edges = [];
// R.reduce((acc, edge) => edge.from === 'in' ? 0 : , {}, edges);

// expecting null

compile(DEFAULT_GRAPH)(new Map(), DEFAULT_GRAPH);