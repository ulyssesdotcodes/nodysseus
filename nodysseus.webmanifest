{
    "name": "Nodysseus",
    "short_name": "nodysseus",
    "description": "Generic node-based editor",
    "start_url": "index.html",
    "display": "fullscreen",
    "icons": [
        {
            "src": "icons/radio-button-off.svg",
            "sizes": "32x32",
            "type": "image/svg"
        }
    ]
}