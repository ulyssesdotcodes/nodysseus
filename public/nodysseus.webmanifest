{
    "name": "Nodysseus",
    "short_name": "nodysseus",
    "description": "Generic node-based editor",
    "start_url": ".",
    "display": "standalone",
    "background_color": "#111",
    "icons": [
        {
            "src": "icons/radio-button-off.svg",
            "sizes": "192x192"
        },
        {
            "src": "icons/radio-button-off.svg",
            "sizes": "512x512"
        }
    ]
}