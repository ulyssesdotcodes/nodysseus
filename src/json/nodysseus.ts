import {default as nodysseusImport} from "./@nodysseus.import.json" assert { type: "json" };

import {default as nodysseusDelete} from "./@nodysseus.delete.json" assert { type: "json" };

export const nodes = [
  nodysseusImport,
  nodysseusDelete
]
