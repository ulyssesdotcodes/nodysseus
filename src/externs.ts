import * as util from "./util"
import {nodysseus_get, resolve_args} from "./nodysseus"
import { ConstRunnable, isNodeRef, isNodeScript, Lib, RefNode, Runnable } from "./types";

 export const create_fn = (runnable: ConstRunnable, lib: Lib) => {
    const graph = util.ancestor_graph(runnable.fn, runnable.graph, lib.data);
  const graphArgs = new Set(Object.values(graph.nodes).filter<RefNode>(isNodeRef).filter(n => n.ref === "arg").map(a => a.value));

   const baseArgs = resolve_args(new Map([...graphArgs].map(a => [a, nodysseus_get(runnable.env, a, lib)])), lib, {});
   const create = ({value: baseArgs}) => {
        graph.id = runnable.fn + "-fn";
        // for(const arg of graphArgs) {
        //   baseArgs[arg] = nodysseus_get(runnable.env, arg, lib)
        // }

        let text = "";
        const _extern_args = {};

        Object.values(graph.nodes).forEach(n => {
          if(isNodeRef(n) && n.ref === "arg") {
            return;
          }
          const noderef = isNodeRef(n) ? lib.data.no.runtime.get_ref(n.ref) : n;
          if(noderef.id === "script") {
            // TODO: extract this logic from node_script
            let inputs = Object.values(graph.edges).filter(e => e.to === n.id).map(edge => ({edge, node: Object.values(graph.nodes).find(n => n.id === edge.from)}));
            text += `function fn_${n.id}(){\n${inputs.map(input => `let ${input.edge.as} = ${isNodeRef(input.node) && input.node.ref === "arg" ? `fnargs["${input.node.value}"] ?? baseArgs["${input.node.value}"]` : `fn_${input.node.id}()`};`).join("\n")}\n\n${isNodeScript(n) ? n.script : n.value}}\n\n`
          } else if(noderef.ref == "extern") {
            _extern_args[n.id] = {};
            const extern = nodysseus_get(lib, noderef.value, lib)
            let inputs = Object.values(graph.edges).filter(e => e.to === n.id).map(edge => ({edge, node: Object.values(graph.nodes).find(n => n.id === edge.from)}));
            const varset = []
            extern.args.map(a => {
              if(a === "__graph_value" || a === "_node") {
                _extern_args[n.id][a] = a === "__graph_value" ? !isNodeScript(n) ? n.value : undefined
                  : "_node" ? n
                  : undefined;
                varset.push(`let ${a} = _extern_args["${n.id}"]["${a}"];`)
              } else if (a === "_node_args") {
                varset.push(`let ${a} = {\n${inputs.map(input => `${input.edge.as}: ${isNodeRef(input.node) && input.node.ref === "arg" ? `(fnargs.${input.node.value} ?? baseArgs.${input.node.value})` : `fn_${input.edge.from}()`}`).join(",\n")}};`);
              } else {
                const input = inputs.find(i => i.edge.as === a);
                if(!input){
                  varset.push(`let ${a} = undefined;`)
                } else {
                  const inputNode = Object.values(graph.nodes).find(n => n.id === input.edge.from);
                  varset.push(`let ${a} = ${isNodeRef(inputNode) && inputNode.ref === "arg" ? `(fnargs.${input.edge.as} ?? baseArgs.${input.edge.as})` : `fn_${input.edge.from}()`};`)
                }
              }
            })
            text += `function fn_${n.id}(){\n${varset.join("\n")}\nreturn (${extern.fn.toString()})(${extern.args.join(", ")})}\n\n`
          } else if (noderef.id === "get"){
            let inputs = Object.values(graph.edges).filter(e => e.to === n.id).map(edge => ({edge, node: Object.values(graph.nodes).find(n => n.id === edge.from)}));
            text += `function fn_${n.id}(){\n${inputs.map(input => `let ${input.edge.as} = ${isNodeRef(input.node) && input.node.ref === "arg" ? input.node.value === '_lib' ? '_lib' : `(fnargs["${input.node.value}"] ?? baseArgs.${input.node.value})` : `fn_${input.node.id}()`};`).join("\n")}\n\nreturn target.${isNodeRef(n) ? n.value : 'undefined'}}\n\n`
          } else {
            text += `function fn_${n.id}(){\nreturn ${noderef.value}}\n\n`
          }
        })

        const fninputs = Object.values(graph.edges).filter(e => e.to === runnable.fn)

        // for now just assumeing everything is an arg of the last node out
        // TODO: tree walk
        text += `return fn_${runnable.fn}()`//({${[...fninputs].map(rinput => `${rinput.as}: fnargs.${graph.nodes.find(n => n.id === rinput.from).value}`).join(",")}})`
        const fn = new Function("fnargs", "baseArgs", "_extern_args", "import_util", "_lib", text);

        return (args={}) => fn(args, baseArgs, _extern_args, util, lib.data);
      }

    return util.wrapPromise(baseArgs).then(create).value
 }

export const now = (scale?: number) => performance.now() * (scale ?? 1)

export const expect = (a: any, b: any, value: string) => {
  if(a === b) return a 
  else {
    console.log('a:')
    console.log(a)
    console.log("b:")
    console.log(b);
    throw new Error(`${value}: Value a does not match value b`)
  }
}
