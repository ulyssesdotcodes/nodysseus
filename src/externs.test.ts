import {describe, expect, test} from "@jest/globals"
import {nolib} from "./nodysseus"
import { create_fn } from "./externs"

describe('create_fn', () => {
  test('parsing a single script', async () => {
    const fn = create_fn({
      fn: 'set',
      graph: {
        id: expect.getState().currentTestName,
        nodes: [
          {id: "set", ref: "script", value: "return 1 + 1"}
        ],
        edges: []
      },
      args: {}
    }, nolib)

    expect(fn()).toEqual(2);
  })

  test('using an argument', async () => {
    const fn = create_fn({
      fn: 'ret',
      graph: {
        id: expect.getState().currentTestName,
        nodes: [
          {id: "inval", ref: "arg", value: "value"},
          {id: "ret", ref: "script", value: "return 1 + inval"}
        ],
        edges: [
          {from: "inval", to: "ret", as: "inval"}
        ]
      },
      args: {}
    }, nolib)

    expect(fn({value: 2})).toEqual(3);
  })

  test('setting a value', async () => {
    const fn = create_fn({
      fn: 'setval',
      graph: {
        id: expect.getState().currentTestName,
        nodes: [
          {id: "inval", ref: "arg", value: "value"},
          {id: "setval", ref: "set", value: "x"},
          {id: "newval", value: "3"},
        ],
        edges: [
          {from: "inval", to: "setval", as: "target"},
          {from: "newval", to: "setval", as: "value"}
        ]
      },
      args: {}
    }, nolib)

    expect(fn({value: {x: 0}}).x).toEqual(3);
  })

})
