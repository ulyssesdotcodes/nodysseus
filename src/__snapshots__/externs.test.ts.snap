// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create_fn snapshot 1`] = `
{
  "children": [
    {
      "children": [
        {
          "dom_type": "text_value",
          "text": undefined,
        },
      ],
      "dom_type": "div",
      "memo": undefined,
      "props": {},
      "value": undefined,
    },
    {
      "children": [
        {
          "dom_type": "text_value",
          "text": "#node-editor-result{z-index: 0;
left: 0;
top: 0;}
target{position: absolute;
z-index: unset;}
#node-info-wrapper, #graph-actions{z-index: 64;}
#node-editor-editor{z-index: 32;
position: relative;}
#node-editor-error{z-index: 4;}",
        },
      ],
      "dom_type": "style",
      "memo": undefined,
      "props": {
        "key": "result-view",
      },
      "value": undefined,
    },
  ],
  "dom_type": "div",
  "memo": undefined,
  "props": {
    "style": undefined,
  },
  "value": undefined,
}
`;
